lib/core/src/restore.rs

```diff
@@ -0,0 +1,1172 @@
+    lbtc_refund_tx_id: Option<HistoryTxId>,
+    btc_server_lockup_tx_id: Option<HistoryTxId>,
+    btc_claim_tx_id: Option<HistoryTxId>,
+}
+impl PartialSwapState for RecoveredOnchainDataChainSend {
+    fn get_partial_state(&self) -> PaymentState {
+        // TODO How to detect TimedOut state?
+        //     TimedOut: This covers the case when the swap state is still Created and the swap fails to reach the
+        //     Pending state in time. The TimedOut state indicates the lockup tx should never be broadcast.
+        match &self.lbtc_user_lockup_tx_id {
+            Some(_) => match &self.btc_claim_tx_id {
+                Some(_) => PaymentState::Complete,
+                None => match &self.lbtc_refund_tx_id {
+                    Some(tx) => match tx.confirmed {
+                        true => PaymentState::Failed,
+                        false => PaymentState::RefundPending,
+                    },
+                    None => PaymentState::Created,
+                },
+            },
+            None => PaymentState::Created,
+        }
+    }
+}
+
+pub(crate) struct RecoveredOnchainDataChainReceive {
+    lbtc_server_lockup_tx_id: Option<HistoryTxId>,
+    lbtc_server_claim_tx_id: Option<HistoryTxId>,
+    btc_user_lockup_tx_id: Option<HistoryTxId>,
+    btc_refund_tx_id: Option<HistoryTxId>,
+}
+impl PartialSwapState for RecoveredOnchainDataChainReceive {
+    fn get_partial_state(&self) -> PaymentState {
+        match &self.btc_user_lockup_tx_id {
+            Some(_) => match &self.lbtc_server_claim_tx_id {
+                Some(_) => PaymentState::Complete,
+                None => match &self.btc_refund_tx_id {
+                    Some(tx) => match tx.confirmed {
+                        true => PaymentState::Failed,
+                        false => PaymentState::RefundPending,
+                    },
+                    None => PaymentState::Created,
+                },
+            },
+            None => PaymentState::Created,
+        }
+    }
+}
+
+pub(crate) struct RecoveredOnchainData {
+    send: HashMap<String, RecoveredOnchainDataSend>,
+    receive: HashMap<String, RecoveredOnchainDataReceive>,
+    chain_send: HashMap<String, RecoveredOnchainDataChainSend>,
+    chain_receive: HashMap<String, RecoveredOnchainDataChainReceive>,
+}
+
+impl LiquidSdk {
+    /// For each swap, recovers data from chain services.
+    ///
+    /// The returned data include txs and the partial swap state. This is a partial state, because
+    /// certain swap states cannot be determined based on initial and onchain data (e.g. [PaymentState::TimedOut])
+    ///
+    /// The caller is expected to merge this data with any other data available, then persist the
+    /// reconstructed swap.
+    pub(crate) async fn recover_from_onchain(&self, tx_map: TxMap) -> Result<RecoveredOnchainData> {
+        // Immutable DB, as fetched from endpoint
+        let imm_db = self.get_swaps_list().await?;
```

Comment:
Ideally this should be given as parameter to this function so we can use it on swaps that are synced in from outside.

Reply:
Done: da39009268bde34d26b81a505873e8f028acaf0b